Update to  proposal #48 (#78)
strimzi/proposals: 048-avoid-broker-restarts-when-in-recovery.md

# Avoid broker restart when in log recovery state

> > Note: This proposal has been updated since the time it was merged.

Reviewer: ```suggestion
> Note: This proposal has been updated since its initial merge.
```

This proposal describes a solution for KafkaRoller's shortcoming with brokers in log recovery.

## Current situation

The following list outlines the high-level logic currently implemented in `KafkaRoller` for rolling pods:
- Take a list of Kafka pods and reorder them based on the readiness.
- If the pod is the controller, roll it last.

> - If pod is not ready before rolling, wait for it to become ready until the operation timeout reaches.

Reviewer: ```suggestion
- If pod is not ready before rolling, wait for it to become ready until the operation timeout is reached.
```

> - If pod is in a stuck state e.g CrashLoopBackOff, force restart the pod.

Reviewer: ```suggestion
- If pod is in a stuck state e.g `CrashLoopBackOff`, force restart the pod.
```

- If pod is unschedulable, fail the reconcilation. 
- If the AdminClient connection to pod failed, force restart the pod. 
- Check if rolling the pod would have an impact on the availability such as causing under replicated partitions.

> - If rolling the pod does affect the availibility, then restart the pod and wait for pod readiness.

Reviewer: A missing not here?

```suggestion
- If rolling the pod does not affect the availibility, then restart the pod and wait for pod readiness. Otherwise consider other pods for rolling first.
```

## Current situation

> The current logic in KafkaRoller from high level:

Reviewer 2: ```suggestion
The following list outlines the high-level logic currently implemented in `KafkaRoller` for rolling pods
```

- Take a list of Kafka pods and reorder them based on the readiness.
- If the pod is the controller, roll it last.
- If pod is not ready before rolling, wait for it to become ready until the operational timeout reaches.
- If pod is in a stuck state e.g CrashLoopBackOff, force restart the pod.
- If pod is unschedulable, fail the reconcilation. 
- If the AdminClient connection to pod failed, force restart the pod. 
- Check if rolling the pod would have an impact on the availability such as causing under replicated partitions.
- If rolling the pod does affect the availibility, then restart the pod and wait for pod readiness.
- If pod does not became ready within the operational timeout (300000ms by default) after restarting, then retry until the maximum attempt is reached. 

The current flow of KafkaRoller
![The current flow of KafkaRoller](images/048-kafka-roller-current-flow.png)

> A Kafka broker can take a long time to become ready while performing log recovery. While broker is in log recovery, AdminClient connection cannot be made to the broker. In this case, KafkaRoller force restarts the broker and this can continue indefinitely, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.

Reviewer 2: ```suggestion
A Kafka broker can take a long time to become ready while performing log recovery. While the broker is in log recovery, AdminClient connection cannot be made to the broker. In this case, KafkaRoller force restarts the broker and this can continue indefinitely, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
```

The following list outlines the high-level logic that will be implemented in `KafkaRoller` for rolling pods:
1. Wait for pod to become ready if it's not in ready state before checking if it can be rolled.
2. If pod does not become ready within the operation timeout, check the broker state.
3. Use the Pod DNS name to connect to the KafkaAgent API endpoint. If it cannot connect to the endpoint (the connection might timeout or refused because KafkaAgent is stuck or running older version), continue with the current behaviour.
4. Use the cluster operator's client certificate and keys for mTLS authentication.
5. Send `GET v1/broker-state` request to the endpoint.
6. If the HTTP return code is not 200, continue with the current behaviour.
7. If the `brokerState` value is `2` (`RECOVERY`):
     - Log the values of `remainingLogsToRecover` and `remainingSegmentsToRecover` from the JSON response in an `INFO` message.
     - Return an exception to repeat from step 1 until the maximum number of attempts is reached.

>      - If the maximum number of attempts has been reached, include log recovery in the reason for failure.

Reviewer: ```suggestion
     - If the maximum number of attempts has been reached then wait the reconciliation including log recovery in the reason for failure.
```

Currently, if a Kafka pod does not become ready within the operational timeout, it checks how many times the pod has been retried to roll. If it hasn't reached the maximum number of attempts (6 by default and set to 10 by KafkaReconcilor), it reschedules the rolling of the pod by appending it to the list. Otherwise, reconciliation fails.

As described in the graph above, the following steps are added after checking the connection to the broker:

1. If the AdminClient connection failed, check the broker state instead of force restarting the pod.

> 2. Use the Pod DNS name to connect to the KafkaAgent API endpoint. If it cannot connect to the endpoint (the connection might timeout or refused because KafkaAgent is stuck or running older version), continue with the current behaviour.

Reviewer 3: Isn't it easier to check if the broker is running but not ready and then check the Agent instead of trying to connect with Admin API first??

Author: Yes that could be simpler. We could check the broker state after the initial wait for readiness times out (https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java#L371). That would mean pod is running but not ready. If broker is in log recovery after the timeout, throw an exception to skip on rest of the checks. Then next time we retry the pod, again would wait for readiness and check the broker state after the timeout. 

Initially I was going to check the broker state here https://github.com/strimzi/strimzi-kafka-operator/blob/main/cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java#L512 but I think what you suggested makes more sense.

3. Use the cluster operator's client certificate and keys for mTLS authentication.
4. Send `GET v1/broker-state` request to the endpoint.
5. If the HTTP return code is not 200, continue with the current behaviour.

> 6. If the `brokerState` value is `2` (`RECOVERY`)

Reviewer 2: ```suggestion
6. If the `brokerState` value is `2` (`RECOVERY`):
```

     - Log the values of `remainingLogsToRecover` and `remainingSegmentsToRecover` from the JSON response in an `INFO` message.
     - Call `awaitReadiness()` which periodically checks and waits for readiness until the operational timeout reaches.

> 7.  If the `brokerState` value is not `2` (`RECOVERY`), continue with the current behaviour of force restarting the pod and waiting for readiness until the operational timeout reaches.

Reviewer 2: ```suggestion
7. If the `brokerState` value is not `2` (`RECOVERY`), continue with the current behaviour of force restarting the pod and waiting for readiness until the operational timeout reaches.
```

> 8. If the operational timeout is reached, continue with the current behaviour of rescheduling the rolling of the pod. This will result in repeating from step 1 if broker is still in log recovery.

Reviewer 2: ```suggestion
8. If the operational timeout is reached, continue with the current behaviour of rescheduling the rolling of the pod. This will result in repeating from step 1 if the broker is still in log recovery.
```
